Hide internal stack traces

Given the following tests
```ts
import { expect } from "earljs"

describe('example', () => {
  it('regular failure', () => {
    expect(123).toEqual(456)
  })

  it('async failure', async () => {
    async function foo() {
       throw new Error('bar')
    }
    await expect(foo()).toBeRejected('baz')
  })
})
```

The output is:
```

  1) example
       regular failure:

      AssertionError: 123 not equal to 456
Hint: value mismatch
      + expected - actual

      -123
      +456
      
      at Control.assert (/home/piotr/Documents/projects/l2beat/node_modules/earljs/dist/Control.js:14:23)
      at toEqual (/home/piotr/Documents/projects/l2beat/node_modules/earljs/dist/validators/toEqual.js:11:17)
      at Expectation.toEqual (/home/piotr/Documents/projects/l2beat/node_modules/earljs/dist/Expectation.js:48:31)
      at Context.<anonymous> (test/x.test.ts:5:17)
      at processImmediate (internal/timers.js:464:21)

  2) example
       async failure:
     AssertionError: Expected to be rejected with "[Error: baz]" but got "Error: bar"
Hint: error
      at Control.assert (/home/piotr/Documents/projects/l2beat/node_modules/earljs/dist/Control.js:14:23)
      at /home/piotr/Documents/projects/l2beat/node_modules/earljs/dist/validators/toBeRejected.js:38:21
      at Generator.throw (<anonymous>)
      at rejected (/home/piotr/Documents/projects/l2beat/node_modules/earljs/dist/validators/toBeRejected.js:6:65)
```

The output could be:
```

  1) example
       regular failure:

      AssertionError: 123 not equal to 456
Hint: value mismatch
      + expected - actual

      -123
      +456
      
      at Context.<anonymous> (test/x.test.ts:5:17)
      at processImmediate (internal/timers.js:464:21)

  2) example
       async failure:
     AssertionError: Expected to be rejected with "[Error: baz]" but got "Error: bar"
Hint: error
      at Context.<anonymous> (test/x.test.ts:12:19)
      at processImmediate (internal/timers.js:464:21)
```

This can be achieved by creating the error in the `Control` constructor and removing `node_modules/earljs` entries from the stack trace thereby preserving both the synchronous stack trace as well as removing internals.

--- a/.changeset/gold-bats-collect.md
+++ b/.changeset/gold-bats-collect.md
@@ -1,0 +1,5 @@
+---
+'earljs': minor
+---
+
+Assertion errors throw with clean (without internal entries) stack traces

--- a/packages/earljs/package.json
+++ b/packages/earljs/package.json
@@ -54,6 +54,7 @@
   "devDependencies": {
     "@microsoft/tsdoc": "^0.13.0",
     "@types/debug": "^4.1.5",
-    "@types/lodash": "^4.14.150"
+    "@types/lodash": "^4.14.150",
+    "error-stack-parser": "^2.0.6"
   }
 }

--- a/packages/earljs/src/Control.ts
+++ b/packages/earljs/src/Control.ts
@@ -4,31 +4,31 @@ import { ValidationResult } from './validators/common'
 
 export class Control<T> {
   public testRunnerCtx = getTestRunnerIntegration()
-  private error = new AssertionError()
+  private stack = AssertionError.getCleanStack()
 
   constructor(public actual: T, public isNegated: boolean, private extraMessage?: string) {}
 
   assert = (result: ValidationResult) => {
     if (this.isNegated === result.success) {
-      this.error.update({
+      throw new AssertionError({
         message: result.success ? result.negatedReason : result.reason,
+        stack: this.stack,
         actual: result.actual,
         expected: result.expected,
         extraMessage: this.extraMessage,
         hint: result.hint,
       })
-      throw this.error
     }
   }
 
   fail = (result: Omit<ValidationResult, 'success' | 'negatedReason'>): never => {
-    this.error.update({
+    throw new AssertionError({
       message: result.reason,
+      stack: this.stack,
       actual: result.actual,
       expected: result.expected,
       extraMessage: this.extraMessage,
       hint: result.hint,
     })
-    throw this.error
   }
 }

--- a/packages/earljs/src/Control.ts
+++ b/packages/earljs/src/Control.ts
@@ -4,28 +4,31 @@ import { ValidationResult } from './validators/common'
 
 export class Control<T> {
   public testRunnerCtx = getTestRunnerIntegration()
+  private error = new AssertionError()
 
   constructor(public actual: T, public isNegated: boolean, private extraMessage?: string) {}
 
   assert = (result: ValidationResult) => {
     if (this.isNegated === result.success) {
-      throw new AssertionError({
+      this.error.update({
         message: result.success ? result.negatedReason : result.reason,
         actual: result.actual,
         expected: result.expected,
         extraMessage: this.extraMessage,
         hint: result.hint,
       })
+      throw this.error
     }
   }
 
   fail = (result: Omit<ValidationResult, 'success' | 'negatedReason'>): never => {
-    throw new AssertionError({
+    this.error.update({
       message: result.reason,
       actual: result.actual,
       expected: result.expected,
       extraMessage: this.extraMessage,
       hint: result.hint,
     })
+    throw this.error
   }
 }

--- a/packages/earljs/src/errors/AssertionError.ts
+++ b/packages/earljs/src/errors/AssertionError.ts
@@ -1,14 +1,23 @@
 import { EOL } from 'os'
 
+interface AssertionErrorOptions {
+  message: string
+  stack: string
+  actual?: string
+  expected?: string
+  extraMessage?: string
+  hint?: string
+}
+
 /**
  * Assertion error containing optional info about actual / expected value which
  * can be used by test runners like Mocha to pretty print.
  */
 export class AssertionError extends Error {
-  public actual: any
-  public expected: any
+  public actual?: string
+  public expected?: string
 
-  update(options: { message: string; actual: any; expected: any; extraMessage?: string; hint?: string }) {
+  constructor(options: AssertionErrorOptions) {
     let message = options.message
     if (options.extraMessage) {
       message += EOL + 'Extra message: ' + options.extraMessage
@@ -16,36 +25,24 @@ export class AssertionError extends Error {
     if (options.hint) {
       message += EOL + 'Hint: ' + options.hint
     }
-    this.message = message
+    super(message)
+    this.name = 'AssertionError'
     this.actual = options.actual
     this.expected = options.expected
-    this.stack = removeInternalEntries(this.stack)
+    this.stack = `${this.name}: ${this.message}\n${options.stack}`
   }
 
-  constructor() {
-    super('')
-    this.name = 'AssertionError'
-  }
-}
+  static getCleanStack() {
+    // .<validator>, .getControl, new Control, .getCleanStack
+    const entriesToRemove = 4
 
-const STACK_ENTRIES_TO_REMOVE = 3
-// at new Control (src/Control.ts)
-// at __ExpectationImplementation.getControl (src/Expectation.ts)
-// at __ExpectationImplementation.<validator> (src/Expectation.ts)
-
-function removeInternalEntries(stack: string | undefined) {
-  const lines = stack?.split('\n') ?? []
-  let stackEntryCount = 0
-  for (let i = lines.length - 1; i >= 0; i--) {
-    if (/^\s+at\s/.test(lines[i])) {
-      stackEntryCount++
-    } else {
-      break
+    const stack = new Error('message').stack
+    if (stack && stack.startsWith('Error: message\n')) {
+      return stack
+        .split('\n')
+        .slice(entriesToRemove + 1)
+        .join('\n')
     }
+    return stack ?? ''
   }
-  if (stackEntryCount <= STACK_ENTRIES_TO_REMOVE) {
-    return stack
-  }
-  lines.splice(lines.length - stackEntryCount, STACK_ENTRIES_TO_REMOVE)
-  return lines.join('\n')
 }

--- a/packages/earljs/src/errors/AssertionError.ts
+++ b/packages/earljs/src/errors/AssertionError.ts
@@ -1,4 +1,3 @@
-import _ from 'lodash'
 import { EOL } from 'os'
 
 /**
@@ -6,31 +5,47 @@ import { EOL } from 'os'
  * can be used by test runners like Mocha to pretty print.
  */
 export class AssertionError extends Error {
-  private readonly actual: any
-  private readonly expected: any
+  public actual: any
+  public expected: any
 
-  constructor({
-    message,
-    actual,
-    expected,
-    extraMessage,
-    hint,
-  }: {
-    message: string
-    actual: any
-    expected: any
-    extraMessage?: string
-    hint?: string
-  }) {
-    const finalMessage = _.compact([
-      message,
-      extraMessage && 'Extra message: ' + extraMessage,
-      hint && 'Hint: ' + hint,
-    ]).join(EOL)
-
-    super(finalMessage)
-    this.name = 'AssertionError'
-    this.actual = actual
-    this.expected = expected
+  update(options: { message: string; actual: any; expected: any; extraMessage?: string; hint?: string }) {
+    let message = options.message
+    if (options.extraMessage) {
+      message += EOL + 'Extra message: ' + options.extraMessage
+    }
+    if (options.hint) {
+      message += EOL + 'Hint: ' + options.hint
+    }
+    this.message = message
+    this.actual = options.actual
+    this.expected = options.expected
+    this.stack = removeInternalEntries(this.stack)
   }
+
+  constructor() {
+    super('')
+    this.name = 'AssertionError'
+  }
+}
+
+const STACK_ENTRIES_TO_REMOVE = 3
+// at new Control (src/Control.ts)
+// at __ExpectationImplementation.getControl (src/Expectation.ts)
+// at __ExpectationImplementation.<validator> (src/Expectation.ts)
+
+function removeInternalEntries(stack: string | undefined) {
+  const lines = stack?.split('\n') ?? []
+  let stackEntryCount = 0
+  for (let i = lines.length - 1; i >= 0; i--) {
+    if (/^\s+at\s/.test(lines[i])) {
+      stackEntryCount++
+    } else {
+      break
+    }
+  }
+  if (stackEntryCount <= STACK_ENTRIES_TO_REMOVE) {
+    return stack
+  }
+  lines.splice(lines.length - stackEntryCount, STACK_ENTRIES_TO_REMOVE)
+  return lines.join('\n')
 }

--- a/packages/earljs/test/errors/stack-traces.test.ts
+++ b/packages/earljs/test/errors/stack-traces.test.ts
@@ -1,0 +1,31 @@
+import { expect } from 'chai'
+import * as errorStackParser from 'error-stack-parser'
+
+import { AssertionError } from '../../src/errors'
+import { expect as earl } from '../../src/expect'
+
+describe('stack traces for errors', () => {
+  it('cleans stack traces for sync errors', () => {
+    try {
+      earl(1).toEqual(2)
+      expect.fail('should throw')
+    } catch (e: any) {
+      expect(e).to.be.instanceOf(AssertionError, 'Earl didnt throw')
+      const stackTrace = errorStackParser.parse(e)
+
+      expect(stackTrace[0].fileName?.endsWith('stack-traces.test.ts')).to.be.true
+    }
+  })
+
+  it('cleans stack traces for async errors', async () => {
+    try {
+      earl(await (async () => 1)()).toEqual(2)
+      expect.fail('should throw')
+    } catch (e: any) {
+      expect(e).to.be.instanceOf(AssertionError, 'Earl didnt throw')
+      const stackTrace = errorStackParser.parse(e)
+
+      expect(stackTrace[0].fileName?.endsWith('stack-traces.test.ts')).to.be.true
+    }
+  })
+})

--- a/packages/earljs/test/validators/snapshots/toMatchSnapshot.test.ts
+++ b/packages/earljs/test/validators/snapshots/toMatchSnapshot.test.ts
@@ -18,16 +18,14 @@ describe('toMatchSnapshot', () => {
     },
   })
 
-  const makeDummyCtrl = (): Control<any> => ({
-    actual: 'test123',
-    isNegated: false,
-    assert: spy() as any,
-    fail: spy() as any,
-    testRunnerCtx: makeDummyTestRunnerCtx(),
-  })
+  class DummyControl extends Control<any> {
+    testRunnerCtx = makeDummyTestRunnerCtx()
+    assert = spy() as any
+    fail = spy() as any
+  }
 
   it('creates new snapshots', () => {
-    const dummyCtrl = makeDummyCtrl()
+    const dummyCtrl = new DummyControl('test123', false)
     const dummyCompareSnapshot: CompareSnapshot = spy(() => {
       return { success: true } as any
     })
@@ -45,7 +43,7 @@ describe('toMatchSnapshot', () => {
   })
 
   it('matches existing snapshots', () => {
-    const dummyCtrl = makeDummyCtrl()
+    const dummyCtrl = new DummyControl('test123', false)
     const dummyCompareSnapshot: CompareSnapshot = spy(() => {
       return { success: false, actual: 'test123', expected: 'abc' } as any
     })

